Fix: Students page now will show all courses enrolled even if it's more than 10.

mikeyarce opened the pull request:
Fixes #5336

### Changes proposed in this Pull Request

Previously, when you loaded the Students page, it would query to get up to 10 courses enrolled by students.  It would show 3, then hide up to 7 which would appear when you clicked on a "more" button.  

The problem with this is that
- If you had more than 10 courses on a student, it would not show them
- It could be a performance issue if you had a page of 20 students all with 10 courses each, and to show all courses could be worse.

What this PR does is
- Moves the "more" button to be an AJAX call that fetches the rest of the courses for the student you are wanting. 
- This way even if a student is enrolled to 100 courses, it happens asynchronously and only one student at a time.

### Testing instructions

* Enable this PR
* Make sure you have at least 11 courses published
* Make sure you have a few students, testing with > 2 is good
* Go to the Students page
* Add all the courses to all the students by clicking the `...` menu like this:
<img width="399" alt="Screen Shot 2022-10-07 at 1 32 33 PM" src="https://user-images.githubusercontent.com/3220162/194648140-6107ee18-2db9-49c5-a6d6-d87599a8e99c.png">
* Try clicking the "+X More" button
* Make sure all the right courses show up for each student.

### Screenshot / Video
![ajax-get-courses3](https://user-images.githubusercontent.com/3220162/194648280-fe7da3bd-aaee-4fe9-a8e7-2ba298f8b839.gif)

mikeyarce committed:
Register AJAX object for script

mikeyarce committed:
Add enrolled courses count query

mikeyarce committed:
Add AJAX action for getting course list HTML

mikeyarce committed:
Update bulk actions view for ajax call

mikeyarce committed:
Update “more” event to get course list html

mikeyarce committed:
Remove console logging

mikeyarce committed:
Sanitize nonce

mikeyarce committed:
Improve error message.

mikeyarce committed:
Fix missing parens

mikeyarce committed:
Improve AJAX error

mikeyarce committed:
Lint fixes

mikeyarce committed:
Fix bulk actions unit test

mikeyarce commented:
For some reason, the PHPUnit test is failing on GitHub even though it passes locally - anyone know why this might be happening?  Is this a cache issue from GitHub Actions?

Edit: Fixed it! ✅

mikeyarce committed:
Fix unit test

dadish commented on assets/js/learners-bulk-actions.js:
```diff
@@ -196,15 +196,41 @@ jQuery( document ).ready( function () {
+							$hiddenPosts.append( data.data );
+						},
+						error: function(errorThrown){
+							$hiddenPosts.append( "<p>There was an error fetching courses: " + errorThrown.statusText + ': ' + errorThrown.status + "</p>" )
```
I know probably nobody will use this, but it's still might be worth to make this translatable.

mikeyarce replied:
👍🏽

dadish commented on includes/class-sensei-learner.php:
```diff
@@ -685,4 +709,41 @@ public static function find_by_query_var( $query_var ) {
+			wp_send_json_error( array( 'error' => esc_html__( 'Insufficient Permissions.', 'sensei-lms' ) ) );
+		}
+		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized -- Don't modify the nonce.
+		if ( ! isset( $_POST['user_id'] ) || ! wp_verify_nonce( ( wp_unslash( $_POST['nonce'] ) ), 'get_course_list' ) ) {
```
Extraneous  parentheses here. 

```suggestion
		if ( ! isset( $_POST['user_id'] ) || ! wp_verify_nonce( wp_unslash( $_POST['nonce'] ), 'get_course_list' ) ) {
```

dadish commented on includes/admin/class-sensei-learners-admin-bulk-actions-view.php:
```diff
@@ -478,7 +480,7 @@ private function get_learner_courses_html( $user_id ) {
 		$hidden_courses  = implode( '', array_slice( $html_items, $visible_count ) );
 
-		return $visible_courses . '<div class="sensei-students__enrolled-courses-detail hidden">' . $hidden_courses . '</div>' . $more_button;
+		return $visible_courses . '<div class="sensei-students__enrolled-courses-detail">' . $hidden_courses . '</div>' . $more_button;
```
Because visible courses and the courses per page are now both `3`, the `$hidden_courses` will always be empty. So we can get rid of it.

dadish reviewed (dismissed):
Good work! Some minor suggestions/comments.

mikeyarce committed:
Add factory method to generate any number of courses

mikeyarce committed:
Add AJAX test for get_course_list

mikeyarce committed:
Remove extra parenthesis

mikeyarce committed:
Move provider outside of loop

mikeyarce committed:
Set user properly

mikeyarce committed:
Remove hidden courses

mikeyarce committed:
Add catch statement

mikeyarce committed:
Make error message translatable

yscik commented on assets/js/learners-bulk-actions.js:
```diff
@@ -196,15 +197,41 @@ jQuery( document ).ready( function () {
+							$hiddenPosts.append( data.data );
+						},
+						error: function(errorThrown){
+							$hiddenPosts.append( __( "<p>There was an error fetching courses: ", 'sensei-lms' ) + errorThrown.statusText + ': ' + errorThrown.status + "</p>" )
```
To keep it consistent, having the opening/closing tags both outside the translatable string:
```suggestion
							$hiddenPosts.append( "<p>" + __( "There was an error fetching courses: ", 'sensei-lms' ) + errorThrown.statusText + ': ' + errorThrown.status + "</p>" )
```

mikeyarce committed:
Keep translation strings consistent

Co-authored-by: Peter Kiss <peter@yscik.com>

mikeyarce commented:
Ready for review if anyone want's to take a look.   A few of the failing tests are to do with the testing library from core, so.... not sure what to do about that.

aaronfc commented on assets/js/learners-bulk-actions.js:
```diff
@@ -196,15 +197,41 @@ jQuery( document ).ready( function () {
+						success:function( data ) {
+							$hiddenPosts.append( data.data );
+						},
+						error: function(errorThrown){
```
Looks like this might need some better formatting? Missing some spaces.

mikeyarce committed:
fix spaces

aaronfc reviewed (dismissed):
LGTM! But there are some tests failing :)

mikeyarce committed:
Fix AJAX unit test

aaronfc committed:
Update with `trunk`.

aaronfc commented:
Hey @mikeyarce, I was checking the failing tests and found something interesting.
I focused on this one:
```
1) Sensei_Learners_Admin_Bulk_Actions_View_Test::testPrepareItems_WhenCalled_ReturnsStudentsWithLastActivityDate
Division by zero

/home/runner/work/sensei/sensei/includes/admin/class-sensei-learners-admin-bulk-actions-view.php:169
/home/runner/work/sensei/sensei/tests/unit-tests/admin/test-class-sensei-learners-admin-bulk-actions-view.php:71
phpvfscomposer:///home/runner/work/sensei/sensei/vendor/phpunit/phpunit/phpunit:52
```

After some debugging (and fighting with my local setup) it looks like the new test you added `Sensei_Learners_Admin_Bulk_Actions_View_AJAX_Test` is affecting the current_screen.
When running `Sensei_Learners_Admin_Bulk_Actions_View_Test` isolated the current_screen is `NULL`, but when running after `Sensei_Learners_Admin_Bulk_Actions_View_AJAX_Test` the current_screen is `front`.

To me this looks like an issue setup/teardown for `WP_Ajax_UnitTestCase` (which, by the way is not happening inn WP 6.1).

I see a couple of alternatives:
1- Check why having `current_screen` set to `front` breaks the test (probably this is our best bet).
2- Try to set the `current_screen` back to `NULL` on the `tearDown` for `Sensei_Learners_Admin_Bulk_Actions_View_AJAX_Test`. Take into account that we would need to do it **after** calling the `parent::tearDown` which is not an ideal approach (we might be breaking things int he future).

Let me know what you think!

mikeyarce committed:
Add override for tearDown tof ix current screen issue

mikeyarce committed:
Merge branch 'fix/students-course-limit-get' of https://github.com/Automattic/sensei into fix/students-course-limit-get

mikeyarce committed:
Remove error reporting

mikeyarce commented:
Thanks @aaronfc 🎉 

Turns out this was a bug with tests before 5.9:  https://core.trac.wordpress.org/ticket/53431

What I did was just override the `tearDown` method for that one Unit test.  There's still one test that fails but that's because the parent method doesn't support PHP 8.0 yet.  

Let me know what you think!

mikeyarce committed:
Merge branch 'trunk' into fix/students-course-limit-get

mikeyarce committed:
Remove tearDown override

mikeyarce committed:
Set current screen to NULL

This is done so that `prepare_items` runs in the right context. Sometimes the current screen will be set to `front` which will break this test.

aaronfc reviewed (dismissed):
Hey @mikeyarce, it looks good!

I was wondering if maybe it would be even easier to just skip the test for <5.9, adding something like:
```
if ( ! version_compare( get_bloginfo( 'version' ), '5.9', '<' ) ) {
	$this->markTestSkipped( 'Test fails with versions prior to 5.9 due to https://core.trac.wordpress.org/ticket/53431.' );
}
```
But I am ok with your proposal too, since that way we will have greater coverage 💪 

About the failing testing in PHP 8.0, do you think you could add the fix here in your PR? I am ok also if we do it in a separate review but the fix looks quite safe and quick. Basically replacing [this method here](https://github.com/Automattic/sensei/blob/100a0e4072329d1649669866ad94d7d8a7c78d95/includes/blocks/course-theme/class-template-style.php#L59) with this:
```
	/**
	 * Renders the block.
	 *
	 * @param array  $attributes The block attributes.
	 * @param string $content The block content.
	 *
	 * @access private
	 *
	 * @return string The block HTML.
	 */
	public function render( $attributes, string $content ): string {
		return $content;
	}
```

WDYT?

mikeyarce committed:
Fix unit test for php 8 compatibility

If a parameter with a default value is followed by a required parameter, the default value has no effect. This is deprecated as of PHP 8.0.0 and can generally be resolved by dropping the default value, without a change in functionality

mikeyarce commented:
Thanks @aaronfc 

I've fixed the PHP 8.0! 
Since we no longer support 5.8, I've removed those tests from running so we won't have that issue.

aaronfc reviewed (approved):
This is awesome 🔥 ! Sorry for the long birthing 😅  🚀
